src/charm.py

```diff
@@ -295,6 +311,22 @@ def _on_install(self, event: InstallEvent) -> None:
         # add licenses
         copy_licenses_to_unit()
 
+    def _on_config_changed(self, event: ConfigChangedEvent) -> None:
+        """Listen to changes in application configuration.
+
+        To prevent a user from migrating a cluster, and causing the component to become
+        unresponsive therefore causing a cluster failure, error the component. This prevents it
+        from executing other hooks with a new role.
+        """
+        # TODO in the future (24.04) support migration of components
+        if self.is_role_changed():
+            logger.error(
+                f"cluster migration currently not supported, cannot change from { self.model.config['role']} to {self.role}"
+            )
+            raise ShardingMigrationError(
```

Comment:
just Q: hm, can we decline config change with exception?

Reply:
seems so, we get:
```
ubuntu@ip-172-31-37-213:~/mongodb-operator$ juju status
Model    Controller  Cloud/Region         Version  SLA          Timestamp
default  overlord    localhost/localhost  3.1.5    unsupported  09:29:42Z

App                Version  Status       Scale  Charm    Channel  Rev  Exposed  Message
config-server-one           error            1  mongodb             0  no       hook failed: "config-changed"
```

Reply:
From Mykola: 
> Let me rephrase: can we return an error when the user runs "juju config ..." instead of success? like with actions

Reply:
Yes, that sounds very reasonable, I can do that